# 1 Find the best approximation matrix A1 of rank 2 of the matrix A in the

1. Find the best approximation matrix $A_1$ of rank 2 of the matrix $A$ in the norm $\Vert \cdot \Vert_2$ and find $\Vert A - A_1 \Vert_2$, where $A$

In [ ]:
import numpy as np

A = np.array([[36,20,-14,-32],[-6,-64,10,40],[-24,52,-55,-16]])
print("rank A: ", np.linalg.matrix_rank(A))
U, E, V = np.linalg.svd(A)
print("U:",U)
print("E:",E)
print("V:",V)

rank A:  3
U: [[-0.33333333  0.66666667  0.66666667]
 [ 0.66666667 -0.33333333  0.66666667]
 [-0.66666667 -0.66666667  0.33333333]]
E: [108.  54.  27.]
V: [[-1.77635684e-16 -7.77777778e-01  4.44444444e-01  4.44444444e-01]
 [ 7.77777778e-01  4.46830501e-17  4.44444444e-01 -4.44444444e-01]
 [ 4.44444444e-01 -4.44444444e-01 -7.77777778e-01  7.45631266e-17]
 [ 4.44444444e-01  4.44444444e-01 -3.25528356e-18  7.77777778e-01]]


In [ ]:
E1 = np.diag(np.array([E[0], E[1], 0]))
E1 = np.c_[ E1, np.zeros(3) ]
E1

array([[108.,   0.,   0.,   0.],
       [  0.,  54.,   0.,   0.],
       [  0.,   0.,   0.,   0.]])

In [ ]:
A1 = (U @ E1 @ V)
A1.round(3)

array([[ 28.,  28.,   0., -32.],
       [-14., -56.,  24.,  40.],
       [-28.,  56., -48., -16.]])

In [ ]:
print("Rank A1: ",np.linalg.matrix_rank(A1))
deltaA = A - A1
U, E, V = np.linalg.svd(deltaA)
print("U:",U)
print("E:",np.diag(E).round(2))
print("V:",V)

Rank A1:  2
U: [[-0.66666667 -0.53206476 -0.52197955]
 [-0.66666667  0.73883954  0.09834479]
 [-0.33333333 -0.41354956  0.84726953]]
E: [[27.  0.  0.]
 [ 0.  0.  0.]
 [ 0.  0.  0.]]
V: [[-4.44444444e-01  4.44444444e-01  7.77777778e-01 -3.94745964e-16]
 [ 6.03563630e-01 -3.61238656e-01  5.51315592e-01 -4.48629798e-01]
 [-3.12859396e-01 -7.61247872e-01  2.56221986e-01  5.06923041e-01]
 [ 5.83351591e-01  3.04101047e-01  1.59571739e-01  7.36043704e-01]]


Answer: $\Vert A - A_1 \Vert_2 = 27$

# 2 Estimate the relative error of the approximate solution (1, 1) of the system

2. Estimate the relative error of the approximate solution $(1, 1)$ of the system $AX = b$ in the norms $\vert \cdot \vert_1$ and $\vert \cdot \vert_2$, using the condition number of the matrix $A$, where

In [ ]:
A = np.array([[4.94,  0.02],[4.04, -8.89]])
b = np.array([[4.93],[-5.03]])



Approximate solution $\hat X = (1, 1)^T$

The relative error $\delta X$

By the property of the condition number: $\frac{1}{\varkappa(A)} \delta B \leq \delta X \leq \varkappa(A)\delta B$

Consider $l_1$-norm first.

Condition number of matrix $A$: $\varkappa(A) = \Vert A \Vert \Vert A^{-1} \Vert$

In [ ]:
kappaA = np.linalg.norm(A, 1) * np.linalg.norm(np.linalg.inv(A), 1)
round(kappaA, 2)

2.64

$\varkappa(A) = 2.64$

Let's take an arbitrary approximate vector $\hat B$

In [ ]:
hatB = b.round()
print(hatB)

[[ 5.]
 [-5.]]


Now we calculate $\delta b = \frac{|\Delta b|_1}{|b|}$

In [ ]:
deltaB = np.linalg.norm(b - hatB, 2) / np.linalg.norm(b, 2)
print(round(deltaB, 2))

0.01


Lower boundary

In [ ]:
round(1/kappaA * deltaB, 4)

0.0041

Upper boundary

In [ ]:
round(kappaA * deltaB, 4)


0.0285

So, $0.0041 \leq \delta x \leq 0.0285$

Now with l2 norm:

In [ ]:
kappaA = np.linalg.norm(A, 2) * np.linalg.norm(np.linalg.inv(A), 2)
hatB = b.round()
deltaB = np.linalg.norm(b - hatB, 2) / np.linalg.norm(b, 2)
print(round(1/kappaA * deltaB, 4))
print(round(kappaA * deltaB, 4))

0.0047
0.0247


So, $0.0047 \leq \delta x \leq 0.0247$

# 3 Solve the system approximately and estimate the relative error of the
 Solve the system approximately and estimate the relative error of the solution in the norms $|\cdot|_1$, $|\cdot|_2$, $|\cdot|_\infty$: 
$$\begin{cases}
-5(6+\varepsilon_1)x + 3(3+\varepsilon_2)y = -2 + \varepsilon_3 \\
2x + (2 + \varepsilon_1)y = -4 + \varepsilon_4
\end{cases}$$ 

where the unknown numbers $\varepsilon_j$ satisfy the conditions $\vert \varepsilon_j \vert < 0.05$ for all $j=1,...,4$.

The same system in vector form $AX = B \Leftrightarrow 
\begin{pmatrix}
-5(6+\varepsilon_1) & 3(3+\varepsilon_2)\\
2 & 2 + \varepsilon_1
\end{pmatrix}
\begin{pmatrix}
x \\
y
\end{pmatrix} = 
\begin{pmatrix}
-2 + \varepsilon_3 \\
-4 + \varepsilon_4
\end{pmatrix}$

Because $ \varepsilon_j $ are unknown, we will find the solution over the widest range. Find the values $ \varepsilon_1 $ and $ \varepsilon_2 $ while $ \varkappa (A) $ is the maximum.

In [ ]:
eps = [-0.05, 0.05, 0, 0]
hatA = np.array([[-30,  9 ],[2, 2 ]])
A = np.array([[-30 - 5*eps[0],  9 + 3*eps[1]],[2, 2 +   eps[0]]])
kappaA = np.linalg.norm(A, 1) * np.linalg.norm(np.linalg.inv(A), 1)
print(round(kappaA, 3))

16.184


So, $\varepsilon_1 = -0.05$ и $\varepsilon_2 = 0.05$

Next find  ε3  and  ε4  where  δB  is maximum.

In [ ]:
eps = [-0.05, 0.05, -0.05, -0.05]
hatB = np.array([-2, -4])
B = np.array([-2 + eps[2], -4 + eps[3]])
deltaB = np.linalg.norm(B - hatB, 1) / np.linalg.norm(B, 1)
print(round(deltaB, 3))

0.016


So $\varepsilon_3 = -0.05$ и $\varepsilon_4 = -0.05$

Let's find the relative error of the solution. 

Lower boundaries

In [ ]:
for n in [1, 2, np.inf]:
    kappaA = np.linalg.norm(A, n) * np.linalg.norm(np.linalg.inv(A), n)
    deltaA = np.linalg.norm(A - hatA, n) / np.linalg.norm(A, n)
    deltaB = np.linalg.norm(B - hatB, n) / np.linalg.norm(B, n)
    print('Norm ',n,' : ', round(1/kappaA * (deltaB + deltaA), 4))

By norm  1  :  0.0015
By norm  2  :  0.002
By norm  inf  :  0.0014


Upper boundaries:

In [ ]:
norms = [1, 2, np.inf]
for n in norms:
    kappaA = np.linalg.norm(A, n) * np.linalg.norm(np.linalg.inv(A), n)
    deltaA = np.linalg.norm(A - hatA, n) / np.linalg.norm(A, n)
    deltaB = np.linalg.norm(B - hatB, n) / np.linalg.norm(B, n)
    print('Norm',n,' : ', round(kappaA * (deltaB + deltaA), 4))

Norm 1  :  0.3928
Norm 2  :  0.3176
Norm inf  :  0.3662


# 4 Find the approximate inverse matrix to the matrix A and evaluate the
4. Find the approximate inverse matrix to the matrix $A$ and evaluate the approximation error with respect to the uniform norm $\Vert \cdot \Vert_1$ if the elements of the matrix $A$ are known with an absolute error of $0.01$:

In [ ]:
A = np.array([[ 5, -8], 
              [7,  9]])

Delta = np.array([[0.01, 0.01], 
                  [0.01, 0.01]])

$A^{-1}$:

In [ ]:
invA = np.linalg.inv(A)
print(np.matrix.round(invA, 2))

[[ 0.09  0.08]
 [-0.07  0.05]]


Relative error from above $\delta A^{-1}\leq \varkappa(A)\frac{\Vert \Delta \Vert}{\Vert A \Vert}$

In [ ]:
round(np.linalg.norm(A, 1) * np.linalg.norm(invA, 1) * (np.linalg.norm(Delta, 1)
         / np.linalg.norm(A, 1)), 3)

0.003

# 5. Use simple iteration method for finding the solution of the given linear
5. Use simple iteration method for finding the solution of the given linear system 
$$\begin{cases}
21x + 6y + 1z = 6 \\
5x + 21y + 3z = 4 \\
6x + 7y + 25z = 2.
\end{cases}$$ 
Determine the iteration number after which the approximation error for each coordinate does not exceed $0.01$ and find the corresponding approximate solution. Start with $x=[0,0,0]^T$.

In [ ]:
A = np.array([[21, 6, 1], 
              [5, 21, 3], 
              [6, 7, 25]])

B = np.array([6, 4, 2])

$X_{k+1} = P X_{k} + B$

So:
$$\begin{cases}
21x + 6y + 1z = 6 \\
5x + 21y + 3z = 4 \\
6x + 7y + 25z = 2
\end{cases} \Leftrightarrow
\begin{cases}
21x = - 6y - 1z + 6 \\
21y = - 5x - 3z + 4 \\
25z = - 6x - 7y + 2
\end{cases} \Leftrightarrow
\begin{cases}
x = - 6y/21 - 1z/21 + 6/21 \\
y = - 5x/21 - 3z/21 + 4/21 \\
z = - 6x/25 - 7y/25 + 2/25
\end{cases}$$

So write the matrix $P$ and vector $B$:

In [ ]:
P = np.array([[0, -6/21, -1/21], [-5/21, 0,-3/21],[-6/25, -7/25, 0 ]])

B = np.array([6/21, 4/21, 2/25])

In [ ]:
round(np.linalg.norm(P, 1), 2)

0.57

Because $\Vert P \Vert_1 < 1$ -> system converges

To get accuracy $\varepsilon$: by the convergence theorem, we need $N$ iterations 

$$ N = \log_{\Vert P \Vert}\left( \frac{\varepsilon(1 - \Vert P \Vert)}{\vert B \vert} \right) - 1$$

For $\varepsilon = 0.01$

In [ ]:
import math
normP = np.linalg.norm(P, 1)
normB = np.linalg.norm(B, 1)
round(math.log((0.01 * (1 - normP))/ normB, normP) - 1, 2)

7.52

We need 8 it-s:

In [ ]:
X = np.array([0, 0, 0])
for i in range(8):
    X = P @ X + B
    print(i+1, X.round(4))

1 [0.2857 0.1905 0.08  ]
2 [ 0.2275  0.111  -0.0419]
3 [ 0.256   0.1423 -0.0057]
4 [ 0.2453  0.1303 -0.0213]
5 [ 0.2495  0.1351 -0.0154]
6 [ 0.2478  0.1333 -0.0177]
7 [ 0.2485  0.134  -0.0168]
8 [ 0.2482  0.1337 -0.0172]


# 6. Find the most influential vertex in the graph using the PageRank 
6. Find the most influential vertex in the graph using the PageRank algorithm with damping factor = $1 − \beta = 0.85$, where the graph adjacency matrix is defined as follows

In [ ]:
A = np.array([[1, 0, 0, 0, 1], 
              [1, 1, 0, 0, 1], 
              [0, 1, 1, 0, 1], 
              [0, 1, 1, 1, 0],
              [0, 1, 1, 0, 1]])

Find transition matrix:

In [ ]:
P = np.array([A[0]/2, A[1]/3, A[2]/3, A[3]/3, A[4]/3]).T
print(P)

[[0.5        0.33333333 0.         0.         0.        ]
 [0.         0.33333333 0.33333333 0.33333333 0.33333333]
 [0.         0.         0.33333333 0.33333333 0.33333333]
 [0.         0.         0.         0.33333333 0.        ]
 [0.5        0.33333333 0.33333333 0.         0.33333333]]


Solution will be in form $X = MX$, where $M = (1 - \beta)P + \beta Q$

β=0.15

By definition Q matrix has form:

In [ ]:
Q = np.ones([5, 5])/5
print(Q)

[[0.2 0.2 0.2 0.2 0.2]
 [0.2 0.2 0.2 0.2 0.2]
 [0.2 0.2 0.2 0.2 0.2]
 [0.2 0.2 0.2 0.2 0.2]
 [0.2 0.2 0.2 0.2 0.2]]


So, solution:

In [ ]:
beta = 0.15
M = (1 - beta)*P + beta*Q
X = np.ones(5)/5
for i in range(15):
    X = M@X
    print(X.round(2))

[0.17 0.26 0.2  0.09 0.29]
[0.18 0.26 0.19 0.05 0.31]
[0.18 0.26 0.19 0.05 0.32]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]
[0.18 0.26 0.19 0.04 0.33]


the most influential vertex is 5.

# 7. Find the value $f(A)$ of the function $f(l) = e^{l+1}$, where

In [ ]:
A = np.array([[3, 5, -2], [-4, -12, 5], [-14, -46, 18]]) + 1*np.identity(3)
print(A)

[[  4.   5.  -2.]
 [ -4. -11.   5.]
 [-14. -46.  19.]]


In [ ]:
term = lambda n: np.linalg.matrix_power(A, n)/float(math.factorial(n))
sum([term(n) for n in range(40)])

array([[  363.25771965,  1069.68762202,  -363.25771965],
       [ -726.51543929, -2199.6318548 ,   746.60097622],
       [-2219.71739172, -6739.49432287,  2279.97400249]])

In [ ]:
from scipy.linalg import expm
expm_A = expm(A)
expm_A

array([[  363.25771965,  1069.68762202,  -363.25771965],
       [ -726.51543929, -2199.6318548 ,   746.60097622],
       [-2219.71739172, -6739.49432287,  2279.97400249]])